docs/csharp/versioning.md

```diff
@@ -0,0 +1,132 @@
+This ensures developers utilizing your library are better prepared for breaking changes.
+* Optional Method Arguments: When you make previously optional method arguments compulsory or change their default value then all code that does not supply those arguments will need to be updated.
+> [!NOTE]
+> Making compulsory arguments optional should have very little effect especially if it doesn't change the method's behaviour.
+
+The easier you make it for your users to upgrade to the new version of your library, the more likely that they will upgrade sooner.
+
+### Application Configuration File
+
+As a .NET developer there's a very high chance you've encountered [the `app.config` file](https://msdn.microsoft.com/en-us/library/1fk1t1t0(v=vs.110).aspx) present in most project types.
+This simple configuration file can go a long way into improving the rollout of new updates. You should generally design your libraries in such
+a way that information that is likely to change regularly is stored in the `app.config` file, this way when such information is updated
+the config file of older versions just needs to be replaced with the new one without the need for recompilation of the library.
+
+## Consuming Libraries
+
+As a developer that consumes .NET libraries built by other developers you're most likely aware that a new version of a library might not be fully compatible with your project
+and you might often find yourself having to update your code to work with those changes.
+
+Lucky for you C# and the .NET ecosystem comes with features and techniques that allow us to easily update our app to work with new versions of libraries that might introduce breaking changes.
+
+### Assembly Binding Redirection
+
+You can use the `app.config` file to update the version of a library your app uses. By adding what is called a [*binding redirect*](https://msdn.microsoft.com/en-us/library/7wd6ex19(v=vs.110).aspx) your
+can use the new library version without having to recompile your app. The following example shows how you would update
+your app's `app.config` file to use the `1.0.1` patch version of `ReferencedLibrary` instead of the `1.0.0` version it was originally compiled with.
+
+```xml
+<dependentAssembly>
+    <assemblyIdentity name="ReferencedLibrary" publicKeyToken="32ab4ba45e0a69a1" culture="en-us" />
+    <bindingRedirect oldVersion="1.0.0" newVersion="1.0.1" />
+</dependentAssembly>
+```
+
+> [!NOTE]
+> This approach will only work if the new version of `ReferencedLibrary` is binary compatible with your app.
+> See the [Backwards Compatibility](#backwards-compatibility) section above for changes to look out for when determining compatibility.
+
+### new
+
+You use the `new` modifier to hide inherited members of a base class.
+This makes your derived classes immune to changes made to the inherited members of the base class
```

Comment:
I'd reword this. Instead of "makes your derived classes immune to changes...", I'd say "is one way derived classes can respond to updates in base classes."